lib/mongo/auth/stringprep.rb

```diff
@@ -0,0 +1,69 @@
+require 'mongo/auth/stringprep/tables'
+require 'mongo/error/string_prep'
+
+module Mongo
+  module Auth
+    module StringPrep
+      extend self
+
+      def prepare(data, mappings, prohibited, options = {})
+        out = apply_maps(data, mappings)
+        normalize(out) if options[:normalize]
+        check_prohibited(out, prohibited)
+        check_bidi(out) if options[:bidi]
+        out
+      end
+
+      private
+
+      def apply_maps(data, mappings)
+        out = ''
+
+        data.each_char do |c|
+          out << mapping(c, mappings)
+        end
+
+        out
+      end
+
+      def check_bidi(out)
+        if out.any? { |c| table_contains?(Tables::C8, c) }
+          raise Error::StringPrep(Error::StringPrep::INVALID_BIDIRECTIONAL)
+        end
+
+        if out.any? { |c| table_contains?(Tables::D1, c) }
+          unless table_contains?(Tables::D1, out.first) && table_contains?(Tables::D1, out.last)
+            raise Error::StringPrep(Error::StringPrep::INVALID_BIDIRECTIONAL)
+          end
+        end
+      end
+
+      def check_prohibited(out, prohibited)
+        out.any? do |c|
+          prohibited.any? do |table|
+            if table_contains?(table, c)
+              raise Error::StringPrep(Error::StringPrep::PROHIBITED_CHARACTER)
+            end
+          end
+        end
+      end
+
+      def mapping(c, mappings)
+        m = mappings.find { |m| m.has_key?(c) }
+        return c if m.nil?
```

Comment:
If you want this to be one line, you could do

```ruby
(m && m[c]) || c
```

Reply:
Done.